### PR TITLE
Expose before filter hook in useSettings for injecting block settings in the editor

### PR DESCRIPTION
**Pull request by ingeniumed**
## What?

This is a combined PR from @alecgeatches and I, that works on the feedback provided to us on our [original governance PR](https://github.com/WordPress/gutenberg/pull/43796).

We have added in a client-side hook [in `useSetting` per @mcsf's recommendation](https://github.com/WordPress/gutenberg/issues/41547#issuecomment-1249590021) to provide block settings to the editor before it renders.

## Why?

This PR is meant to help solve [this issue around governance](https://github.com/WordPress/gutenberg/issues/41547). @dabowman has done a great job explaining this within this issue.
 
[Our previous PR](https://github.com/WordPress/gutenberg/pull/43796) had tried to implement a solution to this issue by baking everything into Gutenberg itself. [The feedback that we received on this PR](https://github.com/WordPress/gutenberg/issues/41547#issuecomment-1249590021) was to minimize the changes that we are proposing into Gutenberg, as the use case that we are trying to solve is not something that a majority of users would take advantage of. In addition there are other issues associated with trying to figure out the best UI to drive this feature as well.

As a result, we went back to work to try to come up with a much simpler solution that tries to minimize the changes going into Gutenberg while still allowing us to be able to solve the original issue. This is that PR as it allows us to make a plugin to get this feature in, without imposing too many changes into Gutenberg.

## How?

Creating a new filter `blockEditor.useSetting.before`: This allows the injection of block settings into the block editor before everything is rendered, so we are able to customize the settings available for a block however we want.

## Testing Instructions

We have added in two tests (scaled down from the [previous PR test changes](https://github.com/WordPress/gutenberg/blob/06f84cf33b1556959d3e876ef5b81b2f045c04f1/packages/block-editor/src/components/use-setting/test/index.js)) that test the new hook addition. There's some `select( blockEditorStore )` mocking necessary to unit test `useSetting()`. Hopefully these tests provide some value, but we'd love to know if there's a better way to approach tests here. These can be run via:

```bash
npm run test:unit -- use-setting/test/index.js
```

### CODE REVIEW

File: lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php
Comment (oandregal) [inline lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php]: This stands out as something problematic to me. Consumers should not use the private classes. See https://github.com/WordPress/gutenberg/issues/45171 for a comprehensive list of issues I see with the current implementation of the theme.json APIs. The most relevant here are the sections about creating public API when necessary and do not leak private implementation details.

What's the data structure you need from a plugin point of view? A map from the block name (`core/paragraph`) to its CSS selector (`p`)? There may be an opportunity to make this public differently (creating a public method for this operation).

**Comment by oandregal** [comment]
> we are able to customize the settings available for a block however we want.

Can you clarify the exact goal here? It can be helpful to understand the specific data needs and also to document it for our future selves when we look back into this. I've read through the issues and this is my understanding: you want to control block settings per block instance. For example, be able to provide different settings for different paragraphs depending on whether the paragraph is within other block, and so on. Is this correct?

File: packages/block-editor/src/components/use-setting/index.js
Comment (oandregal) [inline packages/block-editor/src/components/use-setting/index.js]: My understanding is that the goal is to allow consumers to filter the settings almost on a per-block instance. A few things here:

- the only data that provides contextual information about the block instance to be filtered is the "candidates" list, which are the ancestors of that block. If we aim to filter per-instance, what do you think of only passing the `clientId` instead? Gutenberg can't anticipate all data a consumer may need (`blockName`, the ancestors, etc.), so it'd be best to stay flexible and make the filter about its purpose (filter particular block instances).

- if the filter is a short-circuit one: I find it more readable and more future-proof (code after it may change) if we just return earlier.

```js
result = applyFilters( 'blockEditor.useSetting.before', clientId, path );
if ( undefined !== result ) {
  return result;
}
```

- using `path` vs `normalizedPath` in the filter.

The `normalizedPath` is the runtime path (path within the settings object in the store). `path` may be different from the runtime's. For example: both `useSetting( 'typography.customLineHeight' )` and `useSetting( 'typography.lineHeight' )` work. At some point `typography.customLineHeight` was renamed to `typography.lineHeight` but consumers that execute `useSetting` did not have to adapt to this change. `useSetting` absorbed that complexity within.

What does this mean for the filter? Consumers of the filter should not need to adapt if anything changes, either. It sounds that using `path` is safer than `normalizePath`.

File: lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php
Comment (ingeniumed) [reply lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php]: That is exactly what we need! We would be happy to help create that public method for this operation. How could we kick that off/do it?

**Comment by ingeniumed** [comment]
> > we are able to customize the settings available for a block however we want.
> 
> Can you clarify the exact goal here? It can be helpful to understand the specific data needs and also to document it for our future selves when we look back into this. I've read through the issues and this is my understanding: you want to control block settings per block instance. For example, be able to provide different settings for different paragraphs depending on whether the paragraph is within other block, and so on. Is this correct?

That’s correct - we want to control block settings per block instance depending on context. The context is currently limited to just where the block is located, but we intend to expand this to include role-based information as well as other contexts available in the editor settings.  Giving control over the output of useSetting allows us to determine that context in our plugin and provide a custom result.

File: packages/block-editor/src/components/use-setting/index.js
Comment (ingeniumed) [reply packages/block-editor/src/components/use-setting/index.js]: > the only data that provides contextual information about the block instance to be filtered is the "candidates" list, which are the ancestors of that block. If we aim to filter per-instance, what do you think of only passing the clientId instead? Gutenberg can't anticipate all data a consumer may need (blockName, the ancestors, etc.), so it'd be best to stay flexible and make the filter about its purpose (filter particular block instances).

We based the code here very closely on [@mcsf’s original suggestion](https://github.com/WordPress/gutenberg/issues/41547#issuecomment-1249590021). We agree that it’s impossible and unnecessary to pass all possible context to the filter here. It would still be very useful for our purposes to include `blockName` - we use this to quickly short-circuit on the filter side if the block doesn’t have any special context associated with it. We can query for other data we need, but it would be great if we could keep the `blockName` available since `useSetting()` gets called quite often and it’s an easy way to return quickly via a fast lookup. Looking around the codebase a bit we found [`editor.Autocomplete.completers` filter](https://github.com/WordPress/gutenberg/blob/e0464ed/packages/block-editor/src/components/autocomplete/index.js#L50-L54) doing something similar, but often filters pass the whole `blockType`. Just the `blockName` would be very helpful for us to make the filter implementation return ASAP when it’s not needed for a block.

> if the filter is a short-circuit one: I find it more readable and more future-proof (code after it may change) if we just return earlier.
> using path vs normalizedPath in the filter.

Both of these make sense. We can push up a change that swaps the `normalizedPath` for the path, and returns earlier from the filter so it acts as a short-circuit one.

**Comment by oandregal** [comment]
> we intend to expand this to include role-based information as well as other contexts available in the editor settings

If role-based data means enable/disable settings based on the author's role, there may be an alternate path you can use now. Did you take a look at the existing filters introduced in WordPress 6.1 https://make.wordpress.org/core/2022/10/10/filters-for-theme-json-data/ In the server, hooking into `wp_theme_json_data_user` and provide new settings based on the user's role should be enough.

In that vein, I want to share my expectations for the filter introduced in this PR: it should only be used when the consumer wants to have different setting for different _block instances_. This is: a particular paragraph that meets certain criteria should have different settings than the rest of paragraphs. If the consumer wants to configure all block types based on contextual metadata that does not change in the editor (e.g.: user roles), the existing filters for settings or theme.json ones are a better fit.

So, yeah, the more specific the examples you can share, the better. In many cases, there are already ways to do it without adding more code to the editor :)

File: lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php
Comment (oandregal) [reply lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php]: I've noted this conversation in a couple of threads, for us to keep an eye on. https://github.com/WordPress/gutenberg/issues/45171 and https://github.com/WordPress/gutenberg/issues/45194#issuecomment-1300685965

Would you be able to create a separate pull request that does this only thing (create this public API method)? Ideally, it'd take a single block name and will return its selector. It could live in `lib/compat/wordpress-6.2/get-global-styles-and-settings.php`.

File: packages/block-editor/src/components/use-setting/index.js
Comment (oandregal) [reply packages/block-editor/src/components/use-setting/index.js]: I guess most filters apply to block types, hence the block name as prominent. This filter is special because it aims to actually modify the output based on the block instance, hence why I think `clientId` should be part of the API. I don't have anything against passing _extra_ data that may be convenient, given this is a hot path we should optimize.

File: lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php
Comment (ingeniumed) [reply lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php]: Thanks for breaking that down so well @oandregal! I've put up [this PR](https://github.com/WordPress/gutenberg/pull/45505) to expose this public method. 

I'll update this PR to remove the change to `get_blocks_metadata`

File: packages/block-editor/src/components/use-setting/index.js
Comment (ingeniumed) [reply packages/block-editor/src/components/use-setting/index.js]: Absolutely understand that. We shall add `clientId` to the API as well, for extra flexibility.

**Comment by ingeniumed** [comment]
Thanks for breaking down the expectations you have for the filter being introduced in this PR, so we can be sure that we are on the same page!

> If role-based data means enable/disable settings based on the author's role, there may be an alternate path you can use now. Did you take a look at the existing filters introduced in WordPress 6.1 https://make.wordpress.org/core/2022/10/10/filters-for-theme-json-data/ In the server, hooking into `wp_theme_json_data_user` and provide new settings based on the user's role should be enough.

Thank you for that suggestion! That's a good idea for our role use-case. In general, we do want to allow the consumer to have different settings for different _block_instances_. We would also like to consider surrounding context like the before/after block for this, which would also be block instance specific. We'd love your input as we work out other use-cases to understand server-side ways to accomplish our needs, but right now it unlocks a lot of powerful per-instance customization.

File: lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php
Comment (ingeniumed) [reply lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php]: I've removed the change I made to that method!

File: packages/block-editor/src/components/use-setting/index.js
Comment (ingeniumed) [reply packages/block-editor/src/components/use-setting/index.js]: I've: 

- removed `candidates` and add `clientId` to the filter
- used `path` instead of `normalizedPath`
- done an early return if the filter gives back a value for result

**Review by oandregal — APPROVED** [review]
I've pushed a modification for the filter to be the first step, and nothing else to be executed if a match is found. Thanks for your patience and contribution!

**Comment by ingeniumed** [comment]
Thanks for all your feedback and help, it's much appreciated by @alecgeatches and I!

File: packages/block-editor/src/components/use-setting/index.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/use-setting/index.js]:
```suggestion
			// 0. Allow third parties to filter the block's settings at runtime.
```
Nit: Slight wording tweak.

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for all the iteration on this one @ingeniumed & @alecgeatches 👍 

Given the approach of adding the before filter to `useSetting` has already been accepted on the original issue and via @oandregal's review, I think this PR is very close to landing.

✅ Tests for `useSetting` pass
✅ No issues spotted in the code changes or when testing in the editor

Before we merge this PR, it would be great if you could update the PR description first so that it matches the reduced scope of changes.

**Comment by ingeniumed** [comment]
Thanks a lot for the feedback, and for pointing that out. That should indeed have been updated when the scope was reduced. I have updated the PR description to reflect this, lemme know if you think it needs to be changed further!

**Comment by adamziel** [comment]
I drafted a dev note for this one. @oandregal  @ingeniumed  @aaronrobertshaw would you please review it?

----

## Dev Note: Customize Settings For Any Block in WordPress 6.2

WordPress 6.2 introduces a new client-side filter that allows developers to modify block settings before the editor is rendered.

The filter is called `blockEditor.useSetting.before` and can be used in the JavaScript code as follows:

```js
import { addFilter, removeFilter } from '@wordpress/hooks';

/**
 * Restrict the spacing units that can be
 * selected in the column block to pixels:
 */
addFilter(
	'blockEditor.useSetting.before',
	'myPlugin/useSetting.before',
	( settingValue, settingName, clientId, blockName ) => {
		if ( blockName === 'core/column' && settingName === 'spacing.units' ) {
			return [ 'px' ];
		}
		return settingValue;
	}
);
```

With this filter, developers have better control over what block settings are offered to the user, providing greater control over the editing experience in the Gutenberg editor.

**Comment by alecgeatches** [comment]
@adamziel This looks great to me! I also added [a slightly more complex example to this discussion](https://github.com/WordPress/developer-blog-content/discussions/38#discussioncomment-5096753) which provides behavior not possible with just `theme.json` if that's desired. Thank you for writing this up!

**Comment by ndiego** [comment]
@adamziel @alecgeatches I expanded on the draft above with an additional example to show how to modify settings based on a block's location (thanks for the code @alecgeatches). I just wanted to make sure the dev note captures how powerful this filter can be. I am also planning some additional content on the subject. Let me know what you think. Thanks! 

----

## Dev Note: Customize Settings For Any Block in WordPress 6.2

WordPress 6.2 introduces a new client-side filter that allows developers to modify block settings before the editor is rendered.

The filter is called `blockEditor.useSetting.before` and can be used in the JavaScript code as follows:

```js
import { addFilter } from '@wordpress/hooks';

/**
 * Restrict the spacing units that can be
 * selected in the Column block to pixels:
 */
addFilter(
	'blockEditor.useSetting.before',
	'myPlugin/useSetting.before',
	( settingValue, settingName, clientId, blockName ) => {
		if ( blockName === 'core/column' && settingName === 'spacing.units' ) {
			return [ 'px' ];
		}
		return settingValue;
	}
);
```

This example will restrict the available spacing units for the Column block to just pixels. A similar restriction could be applied using [theme.json filters](https://href.li/?https://make.wordpress.org/core/2022/10/10/filters-for-theme-json-data/) or directly in a theme's theme.json file using [block-level settings](https://developer.wordpress.org/block-editor/how-to-guides/curating-the-editor-experience/#limiting-interface-options-with-theme-json).

That said, the `blockEditor.useSetting.before` filter is unique because it also allows developers to modify settings according to the block's location, neighboring blocks, the current user's role, and more. The possibilities for customization are extensive.

In the following example, text color controls are disabled for the Heading block whenever the block is placed inside of a Media & Text block.  

```js
import { addFilter } from '@wordpress/hooks';
import { getBlockName, getBlockParents } from  '@wordpress/block-editor';

/**
 * Disable text color controls on Heading blocks
 * when placed inside of Media & Text blocks.
 */
addFilter(
	'blockEditor.useSetting.before',
	'myPlugin/useSetting.before',
	( settingValue, settingName, clientId, blockName ) => {
		if ( blockName === 'core/heading' ) {
			const blockParents = getBlockParents( clientId, true );
			const inMediaText = blockParents.some( ( ancestorId ) => getBlockName( ancestorId ) === 'core/media-text' );

			if ( inMediaText && settingName === 'color.text' ) {
       				return false;
     			}
		}

		return settingValue;
	}
);
```

The addition of this filter provides developers with better control over what block settings are presented to the user, enabling them to curate the editor experience like never before.

**Comment by bph** [comment]
[Dev Note needs 1st review ](https://make.wordpress.org/core/?p=102919&preview=1&_ppp=fa32cd4805) Thanks @ndiego for posting the draft.

**Pull request merged**
